DRILL-4523: Disallow using loopback address in distributed mode

arina-ielchiieva opened the pull request.

parthchandra commented on exec/java-exec/src/main/java/org/apache/drill/exec/service/ServiceEngine.java:
```diff
@@ -177,6 +180,16 @@ public void run() {
+  /** If we are in distributed mode, disallow using loopback address to register new drillbit
+   *  unless zookeeper is on the same node with drillbit. */
+  private void checkLoopbackAddress(String address) throws DrillbitStartupException, UnknownHostException {
+    String coordAddress = StringUtils.substringBefore(config.getString(ExecConstants.ZK_CONNECTION), ":");
```
At this point, how do you know you're in distributed and not embedded mode?

arina-ielchiieva replied:
Your comment, made me realize that I was wrong assuming that in embedded mode zookeeper address will be localhost. Actually it will be the same as set in config. So I have passed flag to ServiceEngine class so we know in which mode drill. Pull request was updated.

sudheeshkatkam commented on exec/java-exec/src/main/java/org/apache/drill/exec/service/ServiceEngine.java:
```diff
@@ -177,6 +182,18 @@ public void run() {
+  private void checkLoopbackAddress(String address) throws DrillbitStartupException, UnknownHostException {
+    if (isDistributedMode) {
+      String coordAddress = StringUtils.substringBefore(config.getString(ExecConstants.ZK_CONNECTION), ":");
+      if (!InetAddress.getByName(coordAddress).isLoopbackAddress()
```
In distributed mode, no bit should be allowed to register with a loopback address.

If bit A is allowed to register with a loopback address and bit B tries to communicate with bit A, bit B would be sending messages to itself.

arina-ielchiieva replied:
At some point, we wanted to allow loopback only if we have one node only but it's much safer to disallow loopback in distributed mode at all. Updated pull request.

sudheeshkatkam commented on exec/java-exec/src/main/java/org/apache/drill/exec/service/ServiceEngine.java:
```diff
@@ -177,6 +182,13 @@ public void run() {
+  /** If in distributed mode, disallow using loopback address to register new drillbit. */
+  private void checkLoopbackAddress(String address) throws DrillbitStartupException, UnknownHostException {
+    if (isDistributedMode && InetAddress.getByName(address).isLoopbackAddress()) {
+      throw new DrillbitStartupException("Loopback address is not allowed to be registered in distributed mode");
```
How about "Drillbit is disallowed to bind to loopback address in distributed mode."?

arina-ielchiieva replied:
Sounds reasonable. Updated PR.

arina-ielchiieva replied:
Sounds reasonable. Updated PR.

sudheeshkatkam commented:
+1

How did you test this?

arina-ielchiieva commented:
1. Updated host file so hostname points to the localhost address:
a. started Drill in embedded mode -> started successfully.
b. started drillbit in distributed mode -> expected error.
2. Updated host file so hostname points to the actual IP:
a. started Drill in embedded mode -> started successfully.
b. started drillbit in distributed mode -> started successfully.

halevaavi commented:
Where does drill takes the address. I've the following setup in drill-overide.conf :
drill.exec: {
  cluster-id: "drillbits1",
  zk.connect: "172.16.21.16:2181"
}

172.16.21.16 is a single node that runs both zookeeper & drill and still drillbit complains about running on the loopback address. Any idea why ?

arina-ielchiieva commented:
@halevaavi 
you might check your /etc/hosts file. There should be line like this:
172.16.21.16 your.domain.com  your_hostname
